### ISSUE TITLE
Brackets removed in docs - code blocks

### ISSUE THREAD

**Issue by d2m** [dart-lang/dartdoc#558]
E.g. https://github.com/dart-lang/sdk/blob/master/sdk/lib/js/dartium/js_dartium.dart#L31

    main() => context.callMethod('alert', ['Hello from Dart!']);

becomes in `.../doc/api/dart_js/index.html` 

    main() => context.callMethod('alert', 'Hello from Dart!');

**Comment by sethladd** [comment]
Nice, thanks for the report!

**Comment by sethladd** [comment]
Confirmed!

![screen shot 2015-06-05 at 7 22 01 am](https://cloud.githubusercontent.com/assets/5479/8007897/a098c138-0b53-11e5-9616-2b983ab5ec23.png)

**Comment by sethladd** [comment]
Weird. It's not in the HTML document.

```
<p>This snippet has brackets in parameters:</p>
<pre><code class="prettyprint lang-dart">callMe('alert', 'hello from dart');
</code></pre>
```

**Comment by sethladd** [comment]
Ah. The problem is somewhere in how we process doc comments as markdown. Getting closer.

**Comment by sethladd** [comment]
Good news, actually. We were converting doc ref links the hard way. And it leads to bugs, like this :)

Better way: convert the part of _replaceAllLinks that actually resolves doc references into markdown's Resolver (linkResolver) and pass it into the markdownToHtml function from markdown_processor.dart.

The markdown lib was built to delegate resolving bracket links. We should use it, and stop trying to step through the doc comment string character by character :)

**Comment by sethladd** [comment]
Sending over to Keerti, now that the investigation is done :) There's codin' to be coded.

**Comment by sethladd** [comment]
In markdown_processor.dart, you'll see

```
String renderMarkdownToHtml(String text) {
  return md.markdownToHtml(text, inlineSyntaxes: MARKDOWN_SYNTAXES);
}
```

That should be something like:

```
String renderMarkdownToHtml(String text) {
  return md.markdownToHtml(text, linkResolver: A_FUNCTION_LIKE_REPLACE_ALL_LINKS_BUT_WITHOUT_THE_STRING_MANIPULATION, inlineSyntaxes: MARKDOWN_SYNTAXES);
}
```

**Comment by sethladd** [comment]
Here's a PR with tests (which fail right now, but will pass when we fix this)

https://github.com/dart-lang/dartdoc/pull/561

**Comment by sethladd** [comment]
Also, check out the LinkSyntax class from markdown. Note it takes a linkResolver (which it gets from document, which it gets from markdownToHtml).

LinkSyntax contains this comment:

```
    // If we didn't match refLink or inlineLink, then it means there was
    // nothing after the first square bracket, so it isn't a normal markdown
    // link at all. Instead, we allow users of the library to specify a special
    // resolver function ([linkResolver]) that may choose to handle
    // this. Otherwise, it's just treated as plain text.
```

**Comment by keertip** [comment]
#564. Merged in the tests from #561 too.

**Comment by sethladd** [comment]
Yay! Thanks keetri!

On Monday, June 8, 2015, Keerti Parthasarathy <notifications@github.com>
wrote:

> Closed #558 <https://github.com/dart-lang/dartdoc/issues/558>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dart-lang/dartdoc/issues/558#event-325593673>.
>


-- 
Sent from my Motorola Startac
